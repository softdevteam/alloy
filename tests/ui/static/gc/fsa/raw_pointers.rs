#![feature(gc)]
#![feature(negative_impls)]
#![feature(rustc_private)]
#![allow(dead_code)]
#![allow(unused_variables)]
include!{"./auxiliary/types.rs"}

struct S(*mut u8);

impl Drop for S {
    fn drop(&mut self) {
        use_val(self.0);
    }
}

struct T(*mut u8);

unsafe impl Send for T {}
unsafe impl Sync for T {}

fn main() {
    Gc::new(S(std::ptr::null_mut()));
    //~^ ERROR: The drop method for `S` cannot be safely finalized.

    Gc::new(T(std::ptr::null_mut()));
}
