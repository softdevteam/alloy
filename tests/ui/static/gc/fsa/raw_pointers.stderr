error: The drop method for `S` cannot be safely finalized.
  --> $DIR/raw_pointers.rs:22:13
   |
LL |         use_val(self.0);
   |                 ------
   |                 |
   |                 a finalizer cannot safely dereference this `*mut u8`
   |                 because it might not live long enough
   |                 or be safe to use across threads.
...
LL |     Gc::new(S(std::ptr::null_mut()));
   |     --------^^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<S>` here.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           cannot safely dereference raw pointers. If you are sure that this is safe,
           consider wrapping it in a type which implements `Send + Sync`.

error: aborting due to 1 previous error

